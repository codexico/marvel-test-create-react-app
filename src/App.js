import React, { useState } from 'react';

import Home from './pages/Home/Home';
import Hero from './pages/Hero/Hero';

import './App.css';

function App() {
  const [heroID, setHeroID] = useState();

  function handleClick(id) {
    setHeroID(id);
  }

  return (
    <div className="App">
      {heroID ? <Hero heroID={heroID} /> : <Home handleClick={handleClick} />}
    </div>
  );
}

export default App;
