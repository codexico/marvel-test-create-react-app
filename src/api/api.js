import axios from 'axios';

import { handleError } from './api-utils';

const apikey = process.env.REACT_APP_API_KEY;
const path = 'https://gateway.marvel.com:443/v1/public/characters';

export const apiHeroesList = () => {
  const url = `${path}?apikey=${apikey}`;

  return axios.get(url).catch((err) => handleError(err));
};

export const apiHeroByID = (heroID) => {
  const url = `${path}/${heroID}?apikey=${apikey}`;

  return axios.get(url).catch((err) => handleError(err));
};
