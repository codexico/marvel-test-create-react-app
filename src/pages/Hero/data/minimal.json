{
    "code": 200,
    "status": "Ok",
    "copyright": "© 2020 MARVEL",
    "attributionText": "Data provided by Marvel. © 2020 MARVEL",
    "attributionHTML": "<a href=\"http://marvel.com\">Data provided by Marvel. © 2020 MARVEL</a>",
    "etag": "75b6c8824c323dc143d1104b505f3ce854a0927c",
    "data": {
      "offset": 0,
      "limit": 20,
      "total": 1,
      "count": 1,
      "results": [
        {
          "id": 1010356,
          "name": "Black Queen",
          "description": "",
          "modified": "1969-12-31T19:00:00-0500",
          "thumbnail": {
            "path": "http://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available",
            "extension": "jpg"
          },
          "resourceURI": "http://gateway.marvel.com/v1/public/characters/1010356",
          "comics": {
            "available": 2,
            "collectionURI": "http://gateway.marvel.com/v1/public/characters/1010356/comics",
            "items": [
              {
                "resourceURI": "http://gateway.marvel.com/v1/public/comics/7750",
                "name": "Captain America (1968) #369"
              },
              {
                "resourceURI": "http://gateway.marvel.com/v1/public/comics/18136",
                "name": "X-Men Unlimited (1993) #33"
              }
            ],
            "returned": 2
          },
          "series": {
            "available": 2,
            "collectionURI": "http://gateway.marvel.com/v1/public/characters/1010356/series",
            "items": [
              {
                "resourceURI": "http://gateway.marvel.com/v1/public/series/1996",
                "name": "Captain America (1968 - 1996)"
              },
              {
                "resourceURI": "http://gateway.marvel.com/v1/public/series/3637",
                "name": "X-Men Unlimited (1993 - 2003)"
              }
            ],
            "returned": 2
          },
          "stories": {
            "available": 2,
            "collectionURI": "http://gateway.marvel.com/v1/public/characters/1010356/stories",
            "items": [
              {
                "resourceURI": "http://gateway.marvel.com/v1/public/stories/18018",
                "name": "CAPTAIN AMERICA (1968) #369",
                "type": "cover"
              },
              {
                "resourceURI": "http://gateway.marvel.com/v1/public/stories/67737",
                "name": "X-Men Unlimited 33 cover",
                "type": "cover"
              }
            ],
            "returned": 2
          },
          "events": {
            "available": 0,
            "collectionURI": "http://gateway.marvel.com/v1/public/characters/1010356/events",
            "items": [],
            "returned": 0
          },
          "urls": [
            {
              "type": "detail",
              "url": "http://marvel.com/characters/278/black_queen?utm_campaign=apiRef&utm_source=e34a7870997300d416ccd63679bebfe0"
            },
            {
              "type": "comiclink",
              "url": "http://marvel.com/comics/characters/1010356/black_queen?utm_campaign=apiRef&utm_source=e34a7870997300d416ccd63679bebfe0"
            }
          ]
        }
      ]
    }
  }
