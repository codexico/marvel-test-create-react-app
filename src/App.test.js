import React from 'react';
import {
  cleanup,
  render,
  waitForElement,
  fireEvent,
} from '@testing-library/react';

import dataMinimal from './pages/Hero/data/minimal.json';
import data from './components/HeroesList/data/characters_limit20.json';
import dataSearch from './components/Search/data/search.json';

import { apiHeroesList, apiHeroByID, apiHeroSearch } from './api/api';

jest.mock('./api/api');

afterEach(() => {
  apiHeroesList.mockClear();
  apiHeroByID.mockClear();
  cleanup();
});
import App from './App';

test('App renders Home', async () => {
  apiHeroesList.mockResolvedValue({ data: data });
  const { asFragment, container, getByText } = render(<App />);

  const subtitle = await waitForElement(() => getByText('EXPLORE O UNIVERSO'));
  const Header = container.querySelectorAll('.Header');
  const Search = container.querySelectorAll('.Search');
  const HeroesList = container.querySelectorAll('.HeroesList');

  expect(subtitle).toBeInTheDocument();
  expect(Header).toHaveLength(1);
  expect(Search).toHaveLength(1);
  expect(HeroesList).toHaveLength(1);

  expect(asFragment()).toMatchSnapshot();
});

test('App renders Hero', async () => {
  apiHeroesList.mockResolvedValue({ data: data });
  const { asFragment, container, getByText } = render(<App />);

  expect(apiHeroesList).toHaveBeenCalledTimes(1);

  const HeroCard = await waitForElement(() => getByText('Hulk'));

  apiHeroByID.mockResolvedValue({ data: dataMinimal });
  fireEvent.click(HeroCard);

  const hero = await waitForElement(() => getByText('Black Queen'));

  expect(hero).toBeInTheDocument();

  const Hero = container.querySelectorAll('.Hero');
  expect(Hero).toHaveLength(1);

  expect(asFragment()).toMatchSnapshot();
});

test('App handles search', async () => {
  apiHeroesList.mockResolvedValue({ data: data });
  const { asFragment, container, getByPlaceholderText, getByText } = render(
    <App />
  );

  const input = getByPlaceholderText('Procure por herÃ³is');

  fireEvent.change(input, { target: { value: 'Hulk' } });

  apiHeroSearch.mockResolvedValue({ data: dataSearch });
  fireEvent.submit(input);

  const hero = await waitForElement(() => getByText('Hulk'));
  expect(hero).toBeInTheDocument();

  const HeroCard = container.querySelectorAll('.HeroCard');
  expect(HeroCard).toHaveLength(1);

  expect(apiHeroSearch).toHaveBeenCalledTimes(1);
  expect(apiHeroSearch).toHaveBeenCalledWith('Hulk');
  expect(asFragment()).toMatchSnapshot();
});
