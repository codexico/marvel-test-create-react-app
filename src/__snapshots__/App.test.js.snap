// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link renders correctly 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div
      class="Home"
    >
      <div
        class="Header"
      >
        <svg
          title="Marvel - Search heroes"
        >
          logo.svg
        </svg>
        <h2
          class="Header-title"
        >
          EXPLORE O UNIVERSO
        </h2>
        <h3
          class="Header-subtitle"
        >
          Mergulhe no domínio deslumbrante de todos os personagens clássicos que você ama - e aqueles que você descobrirá em breve!
        </h3>
      </div>
      <div
        class="Search"
      >
        <form
          class="Search-form"
        >
          <div
            class="Search-inputGroup"
          >
            <label
              class="Search-label"
              for=""
            >
              <button
                class="Search-inputButton"
              >
                <svg
                  class="Search-inputIcon"
                >
                  search.svg
                </svg>
              </button>
              <input
                class="Search-input"
                placeholder="Procure por heróis"
              />
            </label>
          </div>
        </form>
      </div>
      <div
        class="HeroesList"
      >
        <div
          class="HeroCard"
        >
          <img
            alt="Hulk"
            class="HeroCard-img"
            src="http://i.annihil.us/u/prod/marvel/i/mg/5/a0/538615ca33ab0.jpg"
          />
          <p
            class="HeroCard-name"
          >
            Hulk
          </p>
        </div>
        <div
          class="HeroCard"
        >
          <img
            alt="Mercury"
            class="HeroCard-img"
            src="http://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available.jpg"
          />
          <p
            class="HeroCard-name"
          >
            Mercury
          </p>
        </div>
        <div
          class="HeroCard"
        >
          <img
            alt="Nicolaos"
            class="HeroCard-img"
            src="http://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available.jpg"
          />
          <p
            class="HeroCard-name"
          >
            Nicolaos
          </p>
        </div>
        <div
          class="HeroCard"
        >
          <img
            alt="The Executioner"
            class="HeroCard-img"
            src="http://i.annihil.us/u/prod/marvel/i/mg/e/d0/4ce5a2ab860be.jpg"
          />
          <p
            class="HeroCard-name"
          >
            The Executioner
          </p>
        </div>
        <div
          class="HeroCard"
        >
          <img
            alt="Champions"
            class="HeroCard-img"
            src="http://i.annihil.us/u/prod/marvel/i/mg/8/b0/4c00376da82bb.jpg"
          />
          <p
            class="HeroCard-name"
          >
            Champions
          </p>
        </div>
        <div
          class="HeroCard"
        >
          <img
            alt="Jack Murdock"
            class="HeroCard-img"
            src="http://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available.jpg"
          />
          <p
            class="HeroCard-name"
          >
            Jack Murdock
          </p>
        </div>
        <div
          class="HeroCard"
        >
          <img
            alt="Janus, the Nega-Man"
            class="HeroCard-img"
            src="http://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available.jpg"
          />
          <p
            class="HeroCard-name"
          >
            Janus, the Nega-Man
          </p>
        </div>
        <div
          class="HeroCard"
        >
          <img
            alt="Mordo"
            class="HeroCard-img"
            src="http://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available.jpg"
          />
          <p
            class="HeroCard-name"
          >
            Mordo
          </p>
        </div>
        <div
          class="HeroCard"
        >
          <img
            alt="Mysterio (Francis Klum)"
            class="HeroCard-img"
            src="http://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available.jpg"
          />
          <p
            class="HeroCard-name"
          >
            Mysterio (Francis Klum)
          </p>
        </div>
        <div
          class="HeroCard"
        >
          <img
            alt="Sabretooth (House of M)"
            class="HeroCard-img"
            src="http://i.annihil.us/u/prod/marvel/i/mg/5/80/4c0033dd26d33.jpg"
          />
          <p
            class="HeroCard-name"
          >
            Sabretooth (House of M)
          </p>
        </div>
        <div
          class="HeroCard"
        >
          <img
            alt="Lady Mastermind"
            class="HeroCard-img"
            src="http://i.annihil.us/u/prod/marvel/i/mg/5/d0/4c00326f63d4c.jpg"
          />
          <p
            class="HeroCard-name"
          >
            Lady Mastermind
          </p>
        </div>
        <div
          class="HeroCard"
        >
          <img
            alt="Living Mummy"
            class="HeroCard-img"
            src="http://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available.jpg"
          />
          <p
            class="HeroCard-name"
          >
            Living Mummy
          </p>
        </div>
        <div
          class="HeroCard"
        >
          <img
            alt="Living Tribunal"
            class="HeroCard-img"
            src="http://i.annihil.us/u/prod/marvel/i/mg/2/a0/4c0031062f91f.jpg"
          />
          <p
            class="HeroCard-name"
          >
            Living Tribunal
          </p>
        </div>
        <div
          class="HeroCard"
        >
          <img
            alt="Scorpion (Ultimate)"
            class="HeroCard-img"
            src="http://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available.jpg"
          />
          <p
            class="HeroCard-name"
          >
            Scorpion (Ultimate)
          </p>
        </div>
        <div
          class="HeroCard"
        >
          <img
            alt="Omega the Unknown"
            class="HeroCard-img"
            src="http://i.annihil.us/u/prod/marvel/i/mg/f/10/4ce5a4e4cd15d.jpg"
          />
          <p
            class="HeroCard-name"
          >
            Omega the Unknown
          </p>
        </div>
        <div
          class="HeroCard"
        >
          <img
            alt="Infant Terrible"
            class="HeroCard-img"
            src="http://i.annihil.us/u/prod/marvel/i/mg/2/f0/4c002f7854e02.jpg"
          />
          <p
            class="HeroCard-name"
          >
            Infant Terrible
          </p>
        </div>
        <div
          class="HeroCard"
        >
          <img
            alt="Crimson King"
            class="HeroCard-img"
            src="http://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available.jpg"
          />
          <p
            class="HeroCard-name"
          >
            Crimson King
          </p>
        </div>
        <div
          class="HeroCard"
        >
          <img
            alt="Captain Cross"
            class="HeroCard-img"
            src="http://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available.jpg"
          />
          <p
            class="HeroCard-name"
          >
            Captain Cross
          </p>
        </div>
        <div
          class="HeroCard"
        >
          <img
            alt="War Machine (Ultimate)"
            class="HeroCard-img"
            src="http://i.annihil.us/u/prod/marvel/i/mg/9/a0/4ce5a5647d38b.jpg"
          />
          <p
            class="HeroCard-name"
          >
            War Machine (Ultimate)
          </p>
        </div>
        <div
          class="HeroCard"
        >
          <img
            alt="Patch"
            class="HeroCard-img"
            src="http://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available.jpg"
          />
          <p
            class="HeroCard-name"
          >
            Patch
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
